lib/level/groups.ex

```diff
@@ -152,6 +152,57 @@ defmodule Level.Groups do
     |> Repo.insert()
   end
 
+  @doc """
+  Deletes a group membership.
+  """
+  @spec delete_group_membership(GroupUser.t()) ::
+          {:ok, GroupUser.t()} | {:error, Ecto.Changeset.t()}
+  def delete_group_membership(group_user) do
+    Repo.delete(group_user)
+  end
+
+  @doc """
+  Updates group membership state.
+  """
+  @spec update_group_membership(Group.t(), SpaceUser.t(), String.t()) ::
+          {:ok, GroupUser.t()} | {:error, GroupUser.t(), Ecto.Changeset.t()}
+  def update_group_membership(group, space_user, state) do
+    case {get_group_membership(group, space_user), state} do
```

Comment:
In generally I like case statements when things are simple, but when you start getting nested case statements, and when some of them start having more than 2 clauses, that's usually the heuristic I use to start pulling that out into a private function and using pattern matching instead of one or more of those case statements. Often you can end up with a really nice pipeline that's pretty easy to follow, and it makes mentally chunking up the behavior of these larger case statements a little easier.

For example, you could refactor this code

```elixir
        case delete_group_membership(group_user) do
          {:ok, _} ->
            {:ok, not_subscribed_membership(group.space_id, space_user, group)}

          {:error, changeset} ->
            {:error, group_user, changeset}
        end
```

into

```elixir
group_user
|> delete_group_membership
|> process_group_membership_deletion(group, space_user, group_user)

defp process_group_membership_deletion({:ok, _}, group, space_user, _) do
  {:ok, not_subscribed_membership(group.space_id, space_user, group)}
end
defp process_group_membership_deletion({:error, changeset}), _, _, group_user) do
   {:error, group_user, changeset}
end
```

This might be just a personal thing, but I've found "avoid nested case statements" to be a pretty helpful heuristic 😃

Reply:
Thanks for the feedback here! I agree in general about avoiding nested cases. I played with a few refactorings here. Interestingly, they all resulted in more lines of code (not necessarily bad) but more importantly, to my eye, they were actually harder to parse then the way it currently stands.

Reply:
👍